### DISCUSSION TITLE
MVP or MVVM type implementations for Riverpod

### DISCUSSION THREAD

**Discussion by dyardy** [rrousselGit/riverpod#1828] [Q&A]
I am impressed with Riverpod. However, I am looking for MVP, MVVM-type implementations to help edit forms (data entry pages that need to respond to validation events, enabling/disabling buttons, hiding/showing fields, etc., and responding to events as a user types in areas.

Any guidance or pointers for this type of implementation would be appreciated.

**Comment by roubachof** [comment]
Hey there!
MVVM is hardly doable in Flutter since it's a declarative MVU framework. You rebuild part of the tree instead of binding to a single property. I am myself a XF/MAUI dev, and I also tried to recreate MVVM in flutter :)
However, I think it is easily implementable with cubits/bloc philosophy...

But don't try to do MVVM with riverpod It's a bad idea. Embrace the philosophy. It doesn't mean it needs to be messy.
You can have your domain layer with repositories and domain providers.
And your presentation layer with widget and presentation providers (like filtering, currently selected item, etc...)
I am currently doing a PoC with riverpod to test its completeness, and it will showcase such an architecture.

**Reply by fredgrott** [reply]
then explain BLoC which is in fact MVVM!

**Comment by dyardy** [comment]
I spent a couple of years with XF and will not return.  I am using Flutter now with an architecture based on gorouter/riverpod/repositories/domain layer etc.   I was, however, looking for a pattern around 'forms' (forms with many inputs/validation, etc.) I have yet to find a typical pattern in this space.

**Reply by roubachof** [reply]
just so you know I just published a riverpod app showcasing a provider oriented architecture :)
https://www.sharpnado.com/falotier-riverpod/
